### PR TITLE
feat(lists): adds a hook to filter list parameters in elgg_list_entities()

### PR DESCRIPTION
**Pull request by hypeJunction**
Adds "prepare, list" hook that filters list options, getter and viewer functions

### CODE REVIEW

**Comment by mrclay** [comment]
This is powerful, but introduces the same issues as https://github.com/Elgg/Elgg/pull/6256: All the listing functions and $options in our views become API surface area. E.g if a plugin adds a custom join, a core change might do the same with the same table alias. Just one way it introduces risk into core changes.

May still be worth it!

**Comment by mrclay** [comment]
Plugins can completely hijack arbitrary listings in code they didn't write. Awesome if done with a surgeon's touch, but it's a sharp knife :)

**Comment by hypeJunction** [comment]
We should start using array keys for joins/wheres, but I don't want to worry about all the numerous way plugins can screw things up. 
My motivation for this is to simplify the sorting logic: https://github.com/hypeJunction/hypeLists/blob/master/views/default/lists/users.php

**Comment by hypeJunction** [comment]
I think I am onto something here, so please check the lists API that I have devised for hypeLists: https://github.com/hypeJunction/hypeLists/commit/ae3fce910348d616c9ac8765fea87172f4f7c8eb

File: docs/guides/hooks-list.rst [hunk 1]
```diff
@@ -103,6 +103,14 @@ System hooks
...
    Filters uncached, page-specific configuration data to pass to the client. :ref:`More info <guides/javascript#config>`
 
+**prepare, list**
+   Triggered by ``elgg_list_entities()`` before getter and viewer functions are called. Can be used to alter the options, the getter and the viewer functions.
+   The hook receives and should return an array containing:
+
+    * ``options`` - an array of options passed to the getter and viewer funtions
+    * ``getter`` - getter function
+    * ``viewer`` - viewer function
+
```
Comment (mrclay) [inline docs/guides/hooks-list.rst]: Maybe add a warning that Elgg may change options/getter/viewer in minor versions, and to improve stability handlers may want to verify they're receiving expected values before changing them.

**Comment by mrclay** [comment]
I'd also like to add a hook that receives the list of entities about to be displayed. This would eliminate the need for this hack: https://github.com/Elgg/Elgg/blob/2.x/mod/likes/views/default/likes/before_lists.php

**Comment by hypeJunction** [comment]
I am shelving this for now. Let's see where denormalization takes us first.

**Comment by hypeJunction** [comment]
Closing for now.

**Pull request closed without merging**